pkg/content/content.go

```diff
@@ -0,0 +1,185 @@
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/goharbor/acceleration-service/pkg/config"
+	"github.com/goharbor/acceleration-service/pkg/remote"
+)
+
+var logger = logrus.WithField("module", "content")
+
+// ContentProvider provides necessary image utils, image content
+// store for image conversion.
+type ContentProvider interface {
+	// Pull pulls source image from remote registry by specified reference.
+	Pull(ctx context.Context, ref string) error
+	// Push pushes target image to remote registry by specified reference,
+	// the desc parameter represents the manifest of targe image.
+	Push(ctx context.Context, desc ocispec.Descriptor, ref string) error
+
+	// Write writes a blob to containerd content store, the blob can be an
+	// image manifest, index, layer or config.
+	Write(ctx context.Context, desc ocispec.Descriptor, reader io.Reader) error
+	// Read reads a blob from containerd content store.
+	Read(ctx context.Context, desc ocispec.Descriptor) ([]byte, error)
+
+	// Image gets the source image object.
+	Image() containerd.Image
+	// Snapshotter gets the snapshotter object of containerd.
+	Snapshotter() snapshots.Snapshotter
+}
+
+type LocalContentProvider struct {
+	image       containerd.Image
+	cfg         *config.ProviderConfig
+	snapshotter snapshots.Snapshotter
+	client      *containerd.Client
+}
+
+func NewLocalContentProvider(
+	cfg *config.ProviderConfig, client *containerd.Client, snapshotter snapshots.Snapshotter,
+) (ContentProvider, error) {
+	return &LocalContentProvider{
+		cfg:         cfg,
+		snapshotter: snapshotter,
+		client:      client,
+	}, nil
+}
+
+func (pvd *LocalContentProvider) Pull(ctx context.Context, ref string) error {
+	refUrl, err := url.Parse(fmt.Sprintf("dummy://%s", ref))
+	if err != nil {
+		return errors.Wrap(err, "parse reference of source image")
+	}
+	var resolver remotes.Resolver
+	if auth, ok := pvd.cfg.Source[refUrl.Host]; ok {
+		resolver = remote.NewResolver(auth.Insecure, remote.NewBasicAuthCredFunc(auth.Auth))
+	} else {
+		resolver = remote.NewResolver(false, remote.NewDockerConfigCredFunc())
+	}
+
+	opts := []containerd.RemoteOpt{
+		// TODO: support multi-platform source images.
+		containerd.WithImageHandler(images.HandlerFunc(
+			func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+				if images.IsLayerType(desc.MediaType) {
+					logger.Debugf("pulling layer %s", desc.Digest)
+				}
+				return nil, nil
+			},
+		)),
+		containerd.WithPullUnpack,
+		containerd.WithResolver(resolver),
+	}
+
+	// Pull the source image from remote registry.
+	image, err := pvd.client.Pull(ctx, ref, opts...)
+	if err != nil {
+		return errors.Wrap(err, "pull source image")
+	}
+	pvd.image = image
+
+	return nil
+}
+
+func (pvd *LocalContentProvider) Write(ctx context.Context, desc ocispec.Descriptor, reader io.Reader) error {
+	digest := desc.Digest.String()
+
+	gcLabels := map[string]string{
+		"containerd.io/gc.ref.content": digest,
+	}
```

Comment:
Why does this GC label point to the blob itself?

Reply:
Ok, I will take a look the usage of GC label.

Reply:
Moved GC label to the manifest blob, does the config or layer blob itself need to be added GC label? I haven't fully understood when containerd will reclaim a blob that has just been written.